File: apps/ios/src/playback/runtime/__tests__/playbackRuntime.test.js
import assert from 'node:assert/strict';
import { test } from 'node:test';

import { createPlaybackRuntime } from '../playbackRuntime.js';

function fakePlayer() {
    const listeners = new Map();
    return {
        audioTrack: null,
        subtitleTrack: null,
        currentTime: 0,
        duration: 100,
        playing: false,
        addListener(name, listener) {
            listeners.set(name, listener);
            return { remove: () => listeners.delete(name) };
        },
        replaceAsync: async function replaceAsync(source) {
            this.source = source;
        },
        play() {
            this.playing = true;
        },
        pause() {
            this.playing = false;
        },
        seekBy(seconds) {
            this.currentTime += seconds;
        },
        listeners
    };
}

function acceptedSession(events, mediaSourceId = 'source-one') {
    return {
        context: {
            mediaSourceId,
            audioStreamIndex: 3,
            subtitleStreamIndex: -1
        },
        pipeline: { pipelineId: 'pipeline-one' },
        telemetry: { log: (event, fields) => events.push({ owner: 'telemetry', event, fields }) },
        sessionMonitor: {
            start: async (snapshot) => events.push({ owner: 'sessionMonitor', event: 'start', snapshot }),
            progress: async (snapshot, options) => events.push({ owner: 'sessionMonitor', event: 'progress', snapshot, options }),
            stop: async (snapshot) => events.push({ owner: 'sessionMonitor', event: 'stop', snapshot })
        },
        attemptedSourceIds: [mediaSourceId],
        startSeconds: 5,
        video: { source: { uri: `http://server/${mediaSourceId}.m3u8`, contentType: 'hls' } }
    };
}

function serverTrackMetadata(audioStreamIndex = 3, subtitleStreamIndex = -1) {
    const audioTracks = [
        { label: 'Japanese FLAC', streamIndex: 1, serverResolved: true },
        { label: 'English AAC', streamIndex: 3, serverResolved: true }
    ];
    const subtitleTracks = [
        { label: 'English ASS', streamIndex: 2, serverResolved: true }
    ];
    return {
        audioTracks,
        subtitleTracks,
        selectedAudioTrack: audioTracks.find((track) => track.streamIndex === audioStreamIndex),
        selectedSubtitleTrack: subtitleTracks.find((track) =>
            track.streamIndex === subtitleStreamIndex
        ) ?? null
    };
}

function waitFor(promise) {
    return Promise.race([
        promise,
        new Promise((resolve, reject) => setTimeout(() => reject(new Error('override not requested')), 100))
    ]);
}

test('should connect native events to session monitoring and terminal cleanup', async () => {
    const events = [];
    const snapshots = [];
    const player = fakePlayer();
    const runtime = createPlaybackRuntime({
        negotiationOptions: {},
        negotiate: async () => acceptedSession(events),
        onSnapshot: (snapshot) => snapshots.push(snapshot),
        player
    });

    await runtime.start();
    player.listeners.get('statusChange')({ status: 'readyToPlay' });
    player.listeners.get('playingChange')({ isPlaying: true });
    player.currentTime = 8;
    player.listeners.get('timeUpdate')({ currentTime: 8, bufferedPosition: 20 });
    await runtime.stop();

    assert.equal(player.source.contentType, 'hls');
    assert.equal(player.currentTime, 8);
    assert.equal(events.filter(({ owner }) => owner === 'sessionMonitor').map(({ event }) => event).join(','), 'start,progress,stop');
    assert.equal(snapshots.at(-1).status, 'ended');
    assert.equal(player.listeners.size, 0);
});

test('should expose native player handoff in runtime snapshots', async () => {
    const snapshots = [];
    const player = fakePlayer();
    const runtime = createPlaybackRuntime({
        negotiationOptions: {},
        async negotiate(options) {
            options.onPipelineProgress({ type: 'resolution_started' });
            options.onPipelineProgress({ type: 'resolution_completed' });
            options.onPipelineProgress({ type: 'release_started' });
            options.onPipelineProgress({ type: 'release_completed' });
            return acceptedSession([]);
        },
        onSnapshot: (snapshot) => snapshots.push(snapshot),
        player
    });

    await runtime.start();

    assert.equal(snapshots.at(-1).pipeline.videoVisible, true);
    assert.equal(snapshots.at(-1).pipeline.visible, true);
    assert.equal(snapshots.at(-1).pipeline.stages.at(-1).status, 'active');

    player.listeners.get('playingChange')({ isPlaying: true });

    assert.equal(snapshots.at(-1).pipeline.visible, false);
    assert.equal(snapshots.at(-1).pipeline.stages.at(-1).status, 'complete');
    await runtime.stop();
});

test('should keep advancing HLS when video-track metadata is absent', async () => {
    const events = [];
    const player = fakePlayer();
    const negotiationCalls = [];
    const runtime = createPlaybackRuntime({
        negotiationOptions: {},
        async negotiate(options) {
            negotiationCalls.push(options);
            return acceptedSession(events, `source-${negotiationCalls.length}`);
        },
        onSnapshot: () => {},
        player
    });

    await runtime.start();
    player.listeners.get('sourceLoad')({
        duration: 100,
        availableAudioTracks: [],
        availableSubtitleTracks: [],
        availableVideoTracks: []
    });
    player.listeners.get('playingChange')({ isPlaying: true });
    player.currentTime = 7;
    player.listeners.get('timeUpdate')({ currentTime: 7, bufferedPosition: 20 });
    await new Promise((resolve) => setTimeout(resolve, 0));

    assert.equal(negotiationCalls.length, 1);
    assert.equal(events.filter(({ owner, event }) => owner === 'sessionMonitor' && event === 'start').length, 1);
    await runtime.stop();
});

test('should expose player commands without policy', async () => {
    const player = fakePlayer();
    const runtime = createPlaybackRuntime({
        negotiationOptions: {},
        negotiate: async () => acceptedSession([]),
        onSnapshot: () => {},
        player
    });

    await runtime.start();
    runtime.pause();
    assert.equal(player.playing, false);
    runtime.play();
    assert.equal(player.playing, true);
    runtime.seekBy(15);
    assert.equal(player.currentTime, 20);
    runtime.seekTo(4);
    assert.equal(player.currentTime, 4);
    const audioTrack = { language: 'ja', label: 'Japanese' };
    const subtitleTrack = { language: 'en', label: 'English' };
    runtime.selectAudioTrack(audioTrack);
    runtime.selectSubtitleTrack(subtitleTrack);
    assert.equal(player.audioTrack, audioTrack);
    assert.equal(player.subtitleTrack, subtitleTrack);
    await runtime.stop();
});

test('should complete player startup when media time advances without playing change', async () => {
    const events = [];
    const player = fakePlayer();
    const runtime = createPlaybackRuntime({
        negotiationOptions: {},
        negotiate: async () => acceptedSession(events),
        onSnapshot: () => {},
        player
    });

    await runtime.start();
    player.listeners.get('timeUpdate')({ currentTime: 6, bufferedPosition: 20 });

    assert.equal(runtime.getSnapshot().status, 'playing');
    assert.equal(runtime.getSnapshot().pipeline.visible, false);
    assert.equal(events.filter(({ event }) => event === 'start').length, 1);
    assert.equal(events.filter(({ event }) => event === 'playback_started').length, 1);

    player.listeners.get('playingChange')({ isPlaying: true });

    assert.equal(events.filter(({ event }) => event === 'start').length, 1);
    assert.equal(events.filter(({ event }) => event === 'playback_started').length, 1);
    await runtime.stop();
});

test('should reject a native player failure before requesting another server attempt', async () => {
    const events = [];
    const snapshots = [];
    const recoverySequence = [];
    const player = fakePlayer();
    const negotiationCalls = [];
    let signalRecovery;
    const recoveryStarted = new Promise((resolve) => {
        signalRecovery = resolve;
    });
    const runtime = createPlaybackRuntime({
        negotiationOptions: { startTimeTicks: 0 },
        async negotiate(options) {
            negotiationCalls.push(options);
            if (negotiationCalls.length === 2) {
                recoverySequence.push('negotiate');
                signalRecovery();
            }
            const accepted = acceptedSession(
                events,
                negotiationCalls.length === 1 ? 'source-one' : 'source-two'
            );
            const stop = accepted.sessionMonitor.stop;
            accepted.sessionMonitor.stop = async (snapshot) => {
                recoverySequence.push('stop');
                await stop(snapshot);
            };
            accepted.startSeconds = options.startTimeTicks / 10_000_000;
            return accepted;
        },
        onSnapshot: (snapshot) => snapshots.push(snapshot),
        player
    });

    await runtime.start();
    player.currentTime = 12;
    player.listeners.get('statusChange')({ status: 'error', error: new Error('decode failed') });
    await recoveryStarted;
    await new Promise((resolve) => setTimeout(resolve, 0));

    assert.equal(negotiationCalls.length, 2);
    assert.equal(negotiationCalls[1].excludedSourceIds.has('source-one'), true);
    assert.equal(player.source.uri, 'http://server/source-two.m3u8');
    const failedIndex = snapshots.findIndex((snapshot) =>
        snapshot.pipeline.stages.some(({ status }) => status === 'failed')
    );
    const retryIndex = snapshots.findIndex((snapshot, index) =>
        index > failedIndex && snapshot.pipeline.attempt === 2
    );
    assert.notEqual(failedIndex, -1);
    assert.ok(retryIndex > failedIndex);
    assert.deepEqual(recoverySequence, ['stop', 'negotiate']);
    await runtime.stop();
});

test('should continue to source three when source two fails during recovery', async () => {
    const events = [];
    const player = fakePlayer();
    const sources = Array.from({ length: 10 }, (_, index) => `source-${index + 1}`);
    const negotiationCalls = [];
    let resolveSourceThree;
    const sourceThreeLoaded = new Promise((resolve) => {
        resolveSourceThree = resolve;
    });
    player.replaceAsync = async function replaceAsync(source) {
        this.source = source;
        if (source.uri.endsWith('/source-2.m3u8')) {
            this.listeners.get('statusChange')({
                status: 'error',
                error: new Error('source two failed')
            });
        }
        if (source.uri.endsWith('/source-3.m3u8')) resolveSourceThree();
    };
    const runtime = createPlaybackRuntime({
        negotiationOptions: { startTimeTicks: 0 },
        async negotiate(options) {
            negotiationCalls.push(options);
            return acceptedSession(events, sources[negotiationCalls.length - 1]);
        },
        onSnapshot: () => {},
        player
    });

    await runtime.start();
    player.listeners.get('statusChange')({
        status: 'error',
        error: new Error('source one failed')
    });
    await waitFor(sourceThreeLoaded);

    assert.equal(negotiationCalls.length, 3);
    assert.deepEqual(
        [...negotiationCalls[2].excludedSourceIds],
        ['source-1', 'source-2']
    );
    assert.equal(player.source.uri, 'http://server/source-3.m3u8');
    assert.equal(negotiationCalls.some((_, index) => index > 2), false);
    await runtime.stop();
});

test('should ignore a source resolved after runtime stop', async () => {
    const player = fakePlayer();
    let resolveNegotiation;
    const negotiation = new Promise((resolve) => {
        resolveNegotiation = resolve;
    });
    const runtime = createPlaybackRuntime({
        negotiationOptions: { startTimeTicks: 0 },
        negotiate: () => negotiation,
        onSnapshot: () => {},
        player
    });

    const start = runtime.start();
    await Promise.resolve();
    await runtime.stop();
    resolveNegotiation(acceptedSession([]));
    await start;

    assert.equal(player.source, undefined);
    assert.equal(player.playing, false);
});

test('should restart at the current position for server audio and subtitle overrides', async () => {
    const events = [];
    const player = fakePlayer();
    const negotiationCalls = [];
    let signalAudioOverride;
    let signalSubtitleOverride;
    const audioOverrideRequested = new Promise((resolve) => {
        signalAudioOverride = resolve;
    });
    const subtitleOverrideRequested = new Promise((resolve) => {
        signalSubtitleOverride = resolve;
    });
    const runtime = createPlaybackRuntime({
        negotiationOptions: { startTimeTicks: 0 },
        async negotiate(options) {
            negotiationCalls.push(options);
            if (negotiationCalls.length === 2) signalAudioOverride();
            if (negotiationCalls.length === 3) signalSubtitleOverride();
            const accepted = acceptedSession(events);
            const override = {
                audioStreamIndex: 3,
                subtitleStreamIndex: -1,
                ...options.trackOverride
            };
            accepted.context.audioStreamIndex = override.audioStreamIndex;
            accepted.context.subtitleStreamIndex = override.subtitleStreamIndex;
            accepted.startSeconds = options.startTimeTicks / 10_000_000;
            accepted.trackMetadata = serverTrackMetadata(
                override.audioStreamIndex,
                override.subtitleStreamIndex
            );
            return accepted;
        },
        onSnapshot: () => {},
        player
    });

    await runtime.start();
    player.currentTime = 12;
    const japanese = runtime.getSnapshot().audioTracks.find(({ streamIndex }) => streamIndex === 1);
    runtime.selectAudioTrack(japanese);
    await waitFor(audioOverrideRequested);
    await new Promise((resolve) => setTimeout(resolve, 0));

    assert.deepEqual(negotiationCalls[1].trackOverride, {
        mediaSourceId: 'source-one',
        audioStreamIndex: 1,
        subtitleStreamIndex: -1
    });
    assert.equal(negotiationCalls[1].startTimeTicks, 120_000_000);

    player.currentTime = 14;
    const englishSubtitle = runtime.getSnapshot().subtitleTracks[0];
    runtime.selectSubtitleTrack(englishSubtitle);
    await waitFor(subtitleOverrideRequested);
    await new Promise((resolve) => setTimeout(resolve, 0));

    assert.deepEqual(negotiationCalls[2].trackOverride, {
        mediaSourceId: 'source-one',
        audioStreamIndex: 1,
        subtitleStreamIndex: 2
    });
    assert.equal(negotiationCalls[2].startTimeTicks, 140_000_000);
    assert.equal(runtime.getSnapshot().selectedAudioTrack.label, 'Japanese FLAC');
    assert.equal(runtime.getSnapshot().selectedSubtitleTrack.label, 'English ASS');
    await runtime.stop();
});

File: apps/android-tv/feature/player/src/test/kotlin/app/homeflix/tv/feature/player/PlaybackRuntimeTest.kt
package app.homeflix.tv.feature.player

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.advanceTimeBy
import kotlinx.coroutines.test.runCurrent
import kotlinx.coroutines.test.runTest
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

@OptIn(ExperimentalCoroutinesApi::class)
class PlaybackRuntimeTest {
    @Test
    fun `should load negotiated source and expose track catalog`() =
        runTest {
            val fixture = RuntimeFixture(this)

            fixture.runtime.start()

            val snapshot = fixture.runtime.snapshot.value
            assertEquals(PlaybackStatus.LOADING, snapshot.status)
            assertEquals(
                listOf("English · 5.1", "Norwegian · Stereo"),
                snapshot.tracks.audioTracks.map(PlaybackTrack::label),
            )
            assertEquals(1, snapshot.tracks.selectedAudioTrack?.streamIndex)
            assertEquals(
                "player",
                snapshot.pipeline.stages
                    .single()
                    .id,
            )
            assertEquals(1, fixture.bindings.size)
            assertEquals(12.0, fixture.bindings.single().loadedStartSeconds)
        }

    @Test
    fun `should start playback once and report start`() =
        runTest {
            val fixture = RuntimeFixture(this)
            fixture.runtime.start()
            val binding = fixture.bindings.single()

            binding.callbacks.onPlayingChange(true)
            binding.callbacks.onPlayingChange(true)
            runCurrent()

            val snapshot = fixture.runtime.snapshot.value
            assertEquals(PlaybackStatus.PLAYING, snapshot.status)
            assertFalse(snapshot.pipeline.visible)
            assertEquals(listOf("start"), fixture.sessionGateway.calls.map(String::toString))
        }

    @Test
    fun `should emit source parity events after negotiation`() =
        runTest {
            val fixture = RuntimeFixture(this)

            fixture.runtime.start()
            runCurrent()

            val events =
                fixture.telemetryEvents.map { event ->
                    event.getValue("event").jsonPrimitive.content
                }
            val parityEvents = listOf("sources_loaded", "source_selected", "tracks_resolved", "source_accepted")
            assertTrue(events.containsAll(parityEvents))
        }

    @Test
    fun `should report pause time in seconds`() =
        runTest {
            val fixture = RuntimeFixture(this)
            fixture.runtime.start()
            val binding = fixture.bindings.single()
            binding.callbacks.onPlayingChange(true)
            binding.callbacks.onTimeUpdate(
                positionSeconds = 25.0,
                durationSeconds = 100.0,
                bufferedSeconds = 30.0,
                playbackAdvanced = true,
            )
            binding.callbacks.onPlayingChange(false)
            runCurrent()

            val pause =
                fixture.telemetryEvents.single {
                    it["event"]?.jsonPrimitive?.content == "playback_paused"
                }
            val positionSeconds =
                pause
                    .getValue("videoCurrentTime")
                    .jsonPrimitive.content
                    .toDouble()
            assertEquals(25.0, positionSeconds)
        }

    @Test
    fun `should report buffering duration`() =
        runTest {
            val fixture = RuntimeFixture(this)
            fixture.runtime.start()
            val callbacks = fixture.bindings.single().callbacks
            fixture.nowMs = 1_000
            callbacks.onPlaybackStateChange(PlaybackEngineState.BUFFERING, playWhenReady = true)
            fixture.nowMs = 6_000
            callbacks.onPlaybackStateChange(PlaybackEngineState.READY, playWhenReady = true)
            runCurrent()

            val started =
                fixture.telemetryEvents.single {
                    it["event"]?.jsonPrimitive?.content == "buffering_started"
                }
            val recovered =
                fixture.telemetryEvents.single {
                    it["event"]?.jsonPrimitive?.content == "buffering_recovered"
                }
            val positionSeconds =
                started
                    .getValue("videoCurrentTime")
                    .jsonPrimitive.content
                    .toDouble()
            val bufferingDurationMs = recovered.getValue("bufferingDurationMs").jsonPrimitive.int
            assertEquals(12.0, positionSeconds)
            assertEquals(5_000, bufferingDurationMs)
        }

    @Test
    fun `should sample memory once per bounded interval`() =
        runTest {
            val fixture =
                RuntimeFixture(
                    scope = this,
                    memoryUsage = {
                        PlaybackMemoryUsage(
                            heapUsedBytes = 90,
                            heapMaxBytes = 192,
                            nativeHeapAllocatedBytes = 20,
                            totalPssKb = 140,
                        )
                    },
                )
            fixture.runtime.start()

            advanceTimeBy(10_000)
            runCurrent()

            val snapshots =
                fixture.telemetryEvents.filter {
                    it["event"]?.jsonPrimitive?.content == "memory_snapshot"
                }
            assertEquals(1, snapshots.size)
            val heapUsedBytes =
                snapshots
                    .single()
                    .getValue("heapUsedBytes")
                    .jsonPrimitive.int
            assertEquals(90, heapUsedBytes)
            fixture.runtime.stop()
        }

    @Test
    fun `should recover on player error excluding failed source`() =
        runTest {
            val fixture = RuntimeFixture(this)
            fixture.runtime.start()
            fixture.bindings
                .single()
                .callbacks
                .onPlayingChange(true)
            runCurrent()

            fixture.bindings
                .single()
                .callbacks
                .onError(
                    PlayerErrorDetails(
                        reason = "ERROR_CODE_DECODING_FAILED",
                        telemetry =
                            mapOf(
                                "errorType" to "MediaCodecVideoDecoderException",
                                "errorCode" to 4_003,
                                "errorName" to "ERROR_CODE_DECODING_FAILED",
                                "errorMessage" to
                                    "renderer=MediaCodecVideoRenderer; decoder=c2.android.hevc.decoder; " +
                                    "diagnostic=android.media.MediaCodec.error_neg_2147483648",
                            ),
                    ),
                )
            runCurrent()

            assertEquals(2, fixture.negotiations.size)
            assertEquals(setOf("source-1"), fixture.negotiations[1].excludedSourceIds)
            assertTrue(fixture.bindings.first().disposed)
            assertEquals(2, fixture.bindings.size)
            assertEquals(2, fixture.runtime.snapshot.value.pipeline.attempt)
            val failureEvent =
                fixture.telemetryEvents.single { event ->
                    event["event"]?.jsonPrimitive?.content == "player_failed"
                }
            assertEquals("MediaCodecVideoDecoderException", failureEvent.getValue("errorType").jsonPrimitive.content)
            assertEquals(4_003, failureEvent.getValue("errorCode").jsonPrimitive.int)
            assertEquals("ERROR_CODE_DECODING_FAILED", failureEvent.getValue("errorName").jsonPrimitive.content)
            assertTrue(
                failureEvent
                    .getValue("errorMessage")
                    .jsonPrimitive.content
                    .contains("c2.android.hevc.decoder"),
            )
            assertEquals(0, failureEvent.getValue("elapsedMs").jsonPrimitive.int)
        }

    @Test
    fun `should recover from backward reset using last confirmed position`() =
        runTest {
            val fixture = RuntimeFixture(this)
            val binding =
                fixture.startPlayingAt(
                    positionSeconds = 600.0,
                    durationSeconds = 1_000.0,
                    bufferedSeconds = 650.0,
                )
            binding.callbacks.onTimeUpdate(
                positionSeconds = 0.0,
                durationSeconds = 0.0,
                bufferedSeconds = 0.0,
                playbackAdvanced = false,
            )
            runCurrent()

            fixture.assertRecoveredAt(startTimeTicks = 6_000_000_000L)
        }

    @Test
    fun `should recover when startup remains buffering`() =
        runTest {
            val fixture = RuntimeFixture(this)
            fixture.runtime.start()
            val binding = fixture.bindings.single()
            binding.callbacks.onPlaybackStateChange(PlaybackEngineState.BUFFERING, playWhenReady = true)
            fixture.nowMs = 30_000
            binding.callbacks.onTimeUpdate(
                positionSeconds = 12.0,
                durationSeconds = 1_000.0,
                bufferedSeconds = 12.0,
                playbackAdvanced = false,
            )
            runCurrent()

            assertEquals(2, fixture.negotiations.size)
            assertEquals(setOf("source-1"), fixture.negotiations.last().excludedSourceIds)
        }

    @Test
    fun `should recover when source ends before expected runtime`() =
        runTest {
            val fixture = RuntimeFixture(this, runTimeTicks = 10_000_000_000L)
            val binding =
                fixture.startPlayingAt(
                    positionSeconds = 600.0,
                    durationSeconds = 1_000.0,
                    bufferedSeconds = 600.0,
                )
            binding.callbacks.onEnded()
            runCurrent()

            fixture.assertRecoveredAt(startTimeTicks = 6_000_000_000L)
        }

    @Test
    fun `should fail when recovery has no source left`() =
        runTest {
            val fixture = RuntimeFixture(this)
            fixture.runtime.start()
            fixture.bindings
                .single()
                .callbacks
                .onPlayingChange(true)
            runCurrent()
            fixture.failNextNegotiation = true

            fixture.bindings
                .single()
                .callbacks
                .onError(PlayerErrorDetails(reason = "decoder failed"))
            runCurrent()

            val snapshot = fixture.runtime.snapshot.value
            assertEquals(PlaybackStatus.FAILED, snapshot.status)
            assertEquals("no compatible playback source", snapshot.reason)
        }

    @Test
    fun `should switch audio natively during direct play`() =
        runTest {
            val fixture = RuntimeFixture(this)
            fixture.runtime.start()
            val binding = fixture.bindings.single()

            fixture.runtime.selectAudioTrack(
                fixture.runtime.snapshot.value.tracks.audioTracks
                    .last(),
            )

            assertEquals(1, fixture.negotiations.size)
            assertEquals(1, binding.nativeAudioSelections.size)
            assertEquals(
                2,
                fixture.runtime.snapshot.value.tracks.selectedAudioTrack
                    ?.streamIndex,
            )
        }

    @Test
    fun `should renegotiate subtitle override when transcoding`() =
        runTest {
            val fixture = RuntimeFixture(this, playMethodOverride = PlayMethod.TRANSCODE)
            fixture.runtime.start()
            runCurrent()

            fixture.runtime.selectSubtitleTrack(
                fixture.runtime.snapshot.value.tracks.subtitleTracks
                    .single(),
            )
            runCurrent()

            assertEquals(2, fixture.negotiations.size)
            val override = fixture.negotiations[1].trackOverride
            assertEquals(3, override?.subtitleStreamIndex)
            assertEquals(1, override?.audioStreamIndex)
        }

    @Test
    fun `should keep pipeline hidden during track renegotiation`() =
        runTest {
            val fixture = RuntimeFixture(this, playMethodOverride = PlayMethod.TRANSCODE)
            fixture.runtime.start()
            fixture.bindings
                .single()
                .callbacks
                .onPlayingChange(true)
            runCurrent()

            fixture.runtime.selectSubtitleTrack(
                fixture.runtime.snapshot.value.tracks.subtitleTracks
                    .single(),
            )
            runCurrent()

            assertEquals(2, fixture.negotiations.size)
            assertFalse(fixture.runtime.snapshot.value.pipeline.visible)
        }

    @Test
    fun `should stop and dispose once`() =
        runTest {
            val fixture = RuntimeFixture(this)
            fixture.runtime.start()
            fixture.bindings
                .single()
                .callbacks
                .onPlayingChange(true)
            runCurrent()

            fixture.runtime.stop()
            fixture.runtime.stop()
            runCurrent()

            assertEquals(PlaybackStatus.ENDED, fixture.runtime.snapshot.value.status)
            assertTrue(fixture.bindings.single().disposed)
            assertEquals(listOf("start", "stop"), fixture.sessionGateway.calls)
            assertNull(fixture.runtime.snapshot.value.reason)
        }
}

private class RuntimeFixture(
    scope: kotlinx.coroutines.CoroutineScope,
    playMethodOverride: PlayMethod = PlayMethod.DIRECT_PLAY,
    runTimeTicks: Long? = null,
    memoryUsage: (() -> PlaybackMemoryUsage)? = null,
) {
    val negotiations = mutableListOf<NegotiationRequest>()
    val bindings = mutableListOf<FakeBinding>()
    val sessionGateway = CallRecordingSessionGateway()
    val telemetryEvents = mutableListOf<kotlinx.serialization.json.JsonObject>()
    var failNextNegotiation = false
    var nowMs = 0L

    private val telemetryGateway =
        object : TelemetryGateway {
            override suspend fun logPipelineEvent(payload: kotlinx.serialization.json.JsonObject) {
                telemetryEvents.add(payload)
            }
        }

    private val pipeline =
        PlaybackPipelineIds(
            itemId = "item-1",
            itemName = "Item",
            pipelineId = "native-abc",
            now = { 0L },
        )

    val runtime =
        PlaybackRuntime(
            scope = scope,
            dependencies =
                PlaybackRuntimeDependencies(
                    negotiate = { request, onEvent ->
                        negotiations.add(request)
                        if (failNextNegotiation) throw NoCompatibleSourceException()
                        onEvent(PipelineEvent.ResolutionStarted)
                        onEvent(PipelineEvent.ReleaseCompleted)
                        accepted(request, playMethodOverride)
                    },
                    bindPlayer = { callbacks -> FakeBinding(callbacks).also(bindings::add) },
                    createReporter = { context ->
                        PlaybackSessionReporter(gateway = sessionGateway, context = context, now = { 0L })
                    },
                    telemetry =
                        PlaybackTelemetry(
                            gateway = telemetryGateway,
                            pipeline = pipeline,
                            scope = scope,
                        ),
                    pipeline = pipeline,
                    monitoring =
                        PlaybackRuntimeMonitoring(
                            nowMs = { nowMs },
                            memoryUsage = memoryUsage,
                        ),
                ),
            request =
                PlaybackStartRequest(
                    item =
                        PlayableItem(
                            id = "item-1",
                            name = "Item",
                            type = "Movie",
                            seriesId = null,
                            seriesName = null,
                            indexNumber = null,
                            parentIndexNumber = null,
                            isMissing = false,
                            resumePositionTicks = 0,
                            backdropUrl = null,
                            runTimeTicks = runTimeTicks,
                        ),
                    userId = "user-1",
                    startTimeTicks = 120_000_000,
                ),
        )

    suspend fun startPlayingAt(
        positionSeconds: Double,
        durationSeconds: Double,
        bufferedSeconds: Double,
    ): FakeBinding {
        runtime.start()
        val binding = bindings.single()
        binding.callbacks.onPlayingChange(true)
        binding.callbacks.onFirstFrame()
        binding.callbacks.onTimeUpdate(
            positionSeconds = positionSeconds,
            durationSeconds = durationSeconds,
            bufferedSeconds = bufferedSeconds,
            playbackAdvanced = true,
        )
        return binding
    }

    fun assertRecoveredAt(startTimeTicks: Long) {
        assertEquals(2, negotiations.size)
        assertEquals(startTimeTicks, negotiations.last().startTimeTicks)
        assertEquals(setOf("source-1"), negotiations.last().excludedSourceIds)
    }

    private fun accepted(
        request: NegotiationRequest,
        playMethod: PlayMethod,
    ): AcceptedPlayback {
        val mediaSource =
            MediaSourceDto(
                id = "source-${negotiations.size}",
                name = "Source",
                supportsDirectPlay = playMethod == PlayMethod.DIRECT_PLAY,
                supportsDirectStream = false,
                supportsTranscoding = true,
                transcodingUrl = "/videos/item-1/main.m3u8",
                transcodingSubProtocol = "hls",
                mediaStreams = fixtureStreams(),
            )
        val released =
            ReleasedPlayback(
                itemId = "item-1",
                mediaSourceId = mediaSource.id,
                playSessionId = "session-1",
                playMethod = playMethod,
                audioStreamIndex = 1,
                subtitleStreamIndex = -1,
                pipelineId = "native-abc",
                attemptId = "native-abc-a${negotiations.size}",
                transcodingUrl = mediaSource.transcodingUrl,
                transcodingSubProtocol = mediaSource.transcodingSubProtocol,
            )
        return AcceptedPlayback(
            released = released,
            mediaSource = mediaSource,
            videoSource = videoSource("http://server", released),
            sourceCount = 3,
            startSeconds = request.startTimeTicks.toDouble() / 10_000_000L,
            videoDelivery = null,
            audioDelivery = null,
            sourceWidth = null,
            sourceHeight = null,
        )
    }
}

private fun fixtureStreams(): List<MediaStreamDto> =
    listOf(
        fixtureStream(index = 1, type = "Audio", language = "eng", channels = 6),
        fixtureStream(index = 2, type = "Audio", language = "nor", channels = 2),
        fixtureStream(index = 3, type = "Subtitle", language = "eng", isExternal = false),
    )

private fun fixtureStream(
    index: Int,
    type: String,
    language: String,
    channels: Int? = null,
    isExternal: Boolean? = null,
): MediaStreamDto =
    MediaStreamDto(
        index = index,
        type = type,
        language = language,
        displayTitle = null,
        title = null,
        channels = channels,
        isForced = false,
        isHearingImpaired = false,
        isExternal = isExternal,
    )

private class FakeBinding(
    val callbacks: PlayerCallbacks,
) : PlayerBinding {
    var loadedStartSeconds: Double? = null
    var disposed = false
    val nativeAudioSelections = mutableListOf<Int>()

    override suspend fun load(
        source: PlayerMediaSource,
        startSeconds: Double,
    ) {
        loadedStartSeconds = startSeconds
    }

    override fun play() = Unit

    override fun pause() = Unit

    override fun seekBy(seconds: Double) = Unit

    override fun seekTo(seconds: Double) = Unit

    override fun selectNativeAudioTrack(typeOrdinal: Int) {
        nativeAudioSelections.add(typeOrdinal)
    }

    override fun selectNativeSubtitleTrack(typeOrdinal: Int?) = Unit

    override fun snapshot(): PlayerSnapshot =
        PlayerSnapshot(positionSeconds = 30.0, durationSeconds = 100.0, isPaused = false)

    override fun dispose() {
        disposed = true
    }
}

private class CallRecordingSessionGateway : SessionReportGateway {
    val calls = mutableListOf<String>()

    override suspend fun reportStart(
        context: SessionContext,
        snapshot: SessionSnapshot,
    ) {
        calls.add("start")
    }

    override suspend fun reportProgress(
        context: SessionContext,
        snapshot: SessionSnapshot,
    ) {
        calls.add("progress")
    }

    override suspend fun reportStop(
        context: SessionContext,
        snapshot: SessionSnapshot,
    ) {
        calls.add("stop")
    }
}
